### PR TITLE
Tpwdcee 6244 feature

### PR DESCRIPTION
**Pull request by ozada**

### CODE REVIEW

File: src/Domain/Exception/FallibleStateException.php [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ * - License can be found under:
+ * https://github.com/wirecard/extension-order-state-module/blob/master/LICENSE
+ */
+
+namespace Wirecard\ExtensionOrderStateModule\Domain\Exception;
+
+/**
+ * Class InvalidPostProcessDataException
+ *
+ * The exception usually occurs if property not provided or value for any property is invalid.
```
Comment (tomazpu) [inline src/Domain/Exception/FallibleStateException.php:15]: In the case when this exception is thrown which is that a post-processing transaction cannot be failed, I think it needs to extend on IgnorableStateException. Also, describe when it happens, what does it mean.
I would name it IgnorablePostProcessingFailure

File: src/Domain/UseCase/PostProcessingPayment/PostProcessingNotification/PartialRefunded.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    /**
+     * @inheritDoc
+     */
+    protected function calculate()
+    {
+        $result = parent::calculate();
+        if ($this->processData->transactionTypeInRange(
+            [Constant::TRANSACTION_TYPE_VOID_PURCHASE, Constant::TRANSACTION_TYPE_REFUND_PURCHASE]
```
Comment (tomazpu) [inline src/Domain/UseCase/PostProcessingPayment/PostProcessingNotification/PartialRefunded.php:37]: As said above I would add the states for TYPE_REFUND_DEBIT, TYPE_CREDIT

File: src/Domain/UseCase/PostProcessingPayment/PostProcessingNotification/Refunded.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    /**
+     * @inheritDoc
+     */
+    protected function calculate()
+    {
+        $result = parent::calculate();
+        if ($this->processData->transactionTypeInRange(
+            [Constant::TRANSACTION_TYPE_VOID_PURCHASE, Constant::TRANSACTION_TYPE_REFUND_PURCHASE]
```
Comment (tomazpu) [inline src/Domain/UseCase/PostProcessingPayment/PostProcessingNotification/Refunded.php:37]: As said above I would add the states for TYPE_REFUND_DEBIT, TYPE_CREDIT

File: tests/unit/Application/Service/OrderStateTest.php [hunk 4]
```diff
@@ -279,4 +285,115 @@ public function testFailingConstructor()
...
+
+        yield "void-purchase_processing_success_pp_notification_refunded" => [
+            Constant::PROCESS_TYPE_POST_PROCESSING_NOTIFICATION,
+            Constant::TRANSACTION_STATE_SUCCESS,
+            Constant::TRANSACTION_TYPE_VOID_PURCHASE,
+            self::EXTERNAL_ORDER_STATE_PROCESSING,
+            100,
+            100,
+            self::EXTERNAL_ORDER_STATE_REFUNDED,
+        ];
```
Comment (tomazpu) [inline tests/unit/Application/Service/OrderStateTest.php:332]: Extending the test than with refund debit and credit.

File: src/Domain/Factory/ProcessFactory.php [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+
+    /**
+     * @return AbstractProcess
+     * @throws InvalidProcessTypeException
+     */
+    public function create()
+    {
+        switch ($this->inputData->getProcessType()) {
```
Comment (flavius) [inline src/Domain/Factory/ProcessFactory.php:51]: TODO: use polymorphism

File: src/Domain/UseCase/PostProcessingPayment/Handler/Notification/PartialRefunded.php [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+    /**
+     * @inheritDoc
+     */
+    protected function calculate()
+    {
+        $result = parent::calculate();
+        if ($this->processData->transactionTypeInRange(
+            [
+                Constant::TRANSACTION_TYPE_VOID_PURCHASE,
```
Comment (flavius) [inline src/Domain/UseCase/PostProcessingPayment/Handler/Notification/PartialRefunded.php:38]: TODO: create common class for this collection of constants.

File: src/Domain/UseCase/PostProcessingPayment/Handler/Notification/Refunded.php [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+    /**
+     * @inheritDoc
+     */
+    protected function calculate()
+    {
+        $result = parent::calculate();
+        if ($this->processData->transactionTypeInRange(
+            [
+                    Constant::TRANSACTION_TYPE_VOID_PURCHASE,
```
Comment (flavius) [inline src/Domain/UseCase/PostProcessingPayment/Handler/Notification/Refunded.php:38]: TODO: create common class for this collection of constants.

File: src/Domain/Registry/AbstractDataRegistry.php [hunk 8]
```diff
@@ -39,11 +39,11 @@ public function __construct()
 
     /**
      * @param string $key
-     * @param object $object
+     * @param object|mixed $object
      * @return AbstractDataRegistry
      * @since 1.0.0
      */
-    protected function attach($key, $object)
+    public function attach($key, $object)
```
Comment (flavius) [inline src/Domain/Registry/AbstractDataRegistry.php:46]: Why was this changed to public?

Comment (ozada) [reply src/Domain/Registry/AbstractDataRegistry.php:46]: 👍
